### PR TITLE
Refactor shopify

### PR DESCRIPTION
**Pull request by saurabh6790**
- [x] Run automated tests
- [ ] Manual testing

### CODE REVIEW

File: erpnext_shopify/api.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+			disable_shopify_sync_on_exception()
+			
+		except requests.exceptions.HTTPError, e:
+			#HTTPError: 402 Client Error: Payment Required 
+			if e.args[0] and e.args[0].startswith("402"):
+				disable_shopify_sync_on_exception()
+				content = _("""Shopify has suspended your account till you complete the payment. We have disabled ERPNext Shopify Sync. Please enable it once your complete the payment at Shopify.""")
+				frappe.sendmail(get_system_managers(), subject=_("Shopify Sync has been disabled"), content=content)
+					
+	elif frappe.local.form_dict.cmd == "erpnext_shopify.erpnext_shopify.doctype.shopify_settings.shopify_settings.sync_shopify":
```
Comment (anandpdoshi) [inline erpnext_shopify/api.py]: You need to change this to `erpnext_shopify.api.sync_shopify`

File: erpnext_shopify/erpnext_shopify/doctype/shopify_error_log/shopify_error_log.json [hunk 2]
```diff
@@ -0,0 +1,125 @@
+{
```
Comment (anandpdoshi) [inline erpnext_shopify/erpnext_shopify/doctype/shopify_error_log/shopify_error_log.json:1]: Why is this needed? Can't we use scheduler log instead?

File: erpnext_shopify/shopify_requests.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+from .exceptions import ShopifyError
+from frappe.utils import get_request_session
+
+def get_shopify_settings():
+	d = frappe.get_doc("Shopify Settings")
+	if d.shopify_url:
+		return d.as_dict()
+	else:
+		frappe.throw(_("Shopify store URL is not configured on Shopify Settings"), ShopifyError)
+
```
Comment (anandpdoshi) [inline erpnext_shopify/shopify_requests.py]: I would prefer naming such that I can use `shopify_requests.get`, `shopify_requests.post`, etc.

File: erpnext_shopify/sync_customers.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	erp_cust = None
+	cust_name = (customer.get("first_name") + " " + (customer.get("last_name") and  customer.get("last_name") or ""))\
+		if customer.get("first_name") else customer.get("email")
+
+	try:
+		erp_cust = frappe.get_doc({
+			"doctype": "Customer",
+			"name": customer.get("id"),
+			"customer_name" : cust_name,
+			"shopify_id": customer.get("id"),
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_customers.py]: preferable fieldname: `shopify_customer_id`

File: erpnext_shopify/sync_customers.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+def sync_customers():
+	sync_shopify_customers()
+	sync_erp_customers()
+
+def sync_shopify_customers():
+	for customer in get_shopify_customers():
+		if not frappe.db.get_value("Customer", {"shopify_id": customer.get('id')}, "name"):
+			create_customer(customer)
+
+def create_customer(customer):
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_customers.py]: variable name should be `shopify_customer`

File: erpnext_shopify/sync_customers.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+			"customer_name" : cust_name,
+			"shopify_id": customer.get("id"),
+			"customer_group": "Commercial",
+			"territory": "All Territories",
+			"customer_type": "Company"
+		}).insert()
+	except:
+		pass
+
+	if erp_cust:
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_customers.py]: use `customer` instead of `erp_cust`

File: erpnext_shopify/sync_customers.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+		create_customer_address(erp_cust, customer)
+
+def create_customer_address(erp_cust, customer):
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_customers.py]: `def create_customer_address(customer, shopify_customer):`

File: erpnext_shopify/sync_customers.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+			"address_line2": address.get("address2"),
+			"city": address.get("city") or "City",
+			"state": address.get("province"),
+			"pincode": address.get("zip"),
+			"country": address.get("country"),
+			"phone": address.get("phone"),
+			"email_id": customer.get("email"),
+			"customer": erp_cust.name,
+			"customer_name":  erp_cust.customer_name
+		}).insert()
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_customers.py:57]: Does a shopify customer address have something we can store like `shopify_address_id`? We will need to sync it.

File: erpnext_shopify/sync_customers.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+def sync_erp_customers():
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_customers.py]: `sync_erpnext_customers`

File: erpnext_shopify/sync_customers.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+		}
+
+		addresses = frappe.db.sql("""select addr.address_line1 as address1, addr.address_line2 as address2,
+			addr.city as city, addr.state as province, addr.country as country, addr.pincode as zip 
+			from tabAddress addr where addr.customer ='%s' """%(customer['customer_name']), as_dict=1)
+
+		if addresses:
+			cust["addresses"] = addresses
+
+		cust = post_request("/admin/customers.json", { "customer": cust})
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_customers.py]: Wont this always create a new customer? What about syncing existing customer data?

File: erpnext_shopify/sync_orders.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+		if valid_customer_and_product(order):
+			create_order(order)
+
+def valid_customer_and_product(order):
+	customer_id = order.get("customer", {}).get("id")
+	if customer_id:
+		if not frappe.db.get_value("Customer", {"shopify_id": customer_id}, "name"):
+			create_customer(order.get("customer"))
+	else:
+		create_log_entry(order, _("Customer is mandatory to create order"))
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py]: Log into scheduler log itself instead of creating a new log

File: erpnext_shopify/sync_orders.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+from .shopify_requests import get_request, get_shopify_orders
+from erpnext.selling.doctype.sales_order.sales_order import make_delivery_note, make_sales_invoice
+
+def sync_orders():
+	sync_shopify_orders()
+
+def sync_shopify_orders():
+	for order in get_shopify_orders():
+		if valid_customer_and_product(order):
+			create_order(order)
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py]: variable name should be `shopify_order`

File: erpnext_shopify/sync_orders.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+		if not frappe.db.get_value("Customer", {"shopify_id": customer_id}, "name"):
+			create_customer(order.get("customer"))
+	else:
+		create_log_entry(order, _("Customer is mandatory to create order"))
+		frappe.msgprint(_("Customer is mandatory to create order"))
+		return False
+	
+	warehouse = frappe.get_doc("Shopify Settings", "Shopify Settings").warehouse
+	for item in order.get("line_items"):
+		if not frappe.db.get_value("Item", {"shopify_id": item.get("product_id")}, "name"):
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py]: Fieldname should be `shopify_product_id` and `shopify_variant_id`

File: erpnext_shopify/sync_orders.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+	shopify_settings = frappe.get_doc("Shopify Settings", "Shopify Settings")
+	so = create_salse_order(order, shopify_settings)
+	if order.get("financial_status") == "paid":
+		create_sales_invoice(order, shopify_settings, so)
+
+	if order.get("fulfillments"):
+		create_delivery_note(order, shopify_settings, so)
+
+def create_salse_order(order, shopify_settings):
+	so = frappe.db.get_value("Sales Order", {"shopify_id": order.get("id")}, "name")
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py]: should be `shopify_order_id`

File: erpnext_shopify/sync_orders.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+def create_order(order):
+	shopify_settings = frappe.get_doc("Shopify Settings", "Shopify Settings")
+	so = create_salse_order(order, shopify_settings)
+	if order.get("financial_status") == "paid":
+		create_sales_invoice(order, shopify_settings, so)
+
+	if order.get("fulfillments"):
+		create_delivery_note(order, shopify_settings, so)
+
+def create_salse_order(order, shopify_settings):
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py]: Typo! `create_sales_order`

File: erpnext_shopify/sync_orders.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+	so = frappe.db.get_value("Sales Order", {"shopify_id": order.get("id")}, "name")
+	if not so:
+		so = frappe.get_doc({
+			"doctype": "Sales Order",
+			"naming_series": shopify_settings.sales_order_series or "SO-Shopify-",
+			"shopify_id": order.get("id"),
+			"customer": frappe.db.get_value("Customer", {"shopify_id": order.get("customer").get("id")}, "name"),
+			"delivery_date": nowdate(),
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py:54]: Can you get the delivery date from the shopify fulfillment?

File: erpnext_shopify/sync_orders.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+			"selling_price_list": shopify_settings.price_list,
+			"ignore_pricing_rule": 1,
+			"apply_discount_on": "Net Total",
+			"discount_amount": get_discounted_amount(order),
+			"items": get_item_line(order.get("line_items"), shopify_settings),
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py]: function names should be `get_order_items` and `get_order_taxes`

File: erpnext_shopify/sync_orders.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+		}).insert()
+
+		so.submit()
+
+	else:
+		so = frappe.get_doc("Sales Order", so)
+
+	return so
+
+def create_sales_invoice(order, shopify_settings, so):
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py]: `order` should be changed to `shopify_order`

File: erpnext_shopify/sync_orders.py [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+	if not frappe.db.get_value("Sales Invoice", {"shopify_id": order.get("id")}, "name") and so.docstatus==1 \
+		and not so.per_billed:
+		si = make_sales_invoice(so.name)
+		si.shopify_id = order.get("id")
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py]: `shopify_order_id`

File: erpnext_shopify/sync_orders.py [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+		si.naming_series = shopify_settings.sales_invoice_series or "SI-Shopify-"
+		si.is_pos = 1
+		si.cash_bank_account = shopify_settings.cash_bank_account
+		si.submit()
+
+def create_delivery_note(order, shopify_settings, so):
+	for fulfillment in order.get("fulfillments"):
+		if not frappe.db.get_value("Delivery Note", {"shopify_id": fulfillment.get("id")}, "name") and so.docstatus==1:
+			dn = make_delivery_note(so.name)
+			dn.shopify_id = fulfillment.get("id")
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py]: `shopify_order_id`

File: erpnext_shopify/sync_orders.py [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+			dn.naming_series = shopify_settings.delivery_note_series or "DN-Shopify-"
+			dn.items = update_items_qty(dn.items, fulfillment.get("line_items"), shopify_settings)
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py]: `get_fulfillment_items`

File: erpnext_shopify/sync_orders.py [hunk 22]
```diff
@@ -0,0 +1,159 @@
...
+			dn.save()
+
+def update_items_qty(dn_items, fulfillment_items, shopify_settings):
+	return [dn_item.update({"qty": item.get("quantity")}) for item in fulfillment_items for dn_item in dn_items\
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py:98]: why not write it as a simple for loop and return dn_items?

File: erpnext_shopify/sync_orders.py [hunk 23]
```diff
@@ -0,0 +1,159 @@
...
+			"item_name": item.get("name"),
+			"rate": item.get("price"),
+			"qty": item.get("quantity"),
+			"stock_uom": item.get("sku"),
+			"warehouse": shopify_settings.warehouse
+		})
+	return items
+
+def get_item_code(item):
+	item_code = frappe.db.get_value("Item", {"shopify_id": item.get("variant_id")}, "item_code")
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py]: `shopify_item` instead of `item`

File: erpnext_shopify/sync_orders.py [hunk 24]
```diff
@@ -0,0 +1,159 @@
...
+
+	return item_code
+
+def get_tax_line(order, shipping_lines, shopify_settings):
+	taxes = []
+	for tax in order.get("tax_lines"):
+		taxes.append({
+			"charge_type": _("On Net Total"),
+			"account_head": get_tax_account_head(tax),
+			"description": tax.get("title") + "-" + cstr(tax.get("rate") * 100.00),
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py]: `"{0} - {1}%".format(tax.get("title"), tax.get("rate") * 100.0)`

File: erpnext_shopify/sync_orders.py [hunk 25]
```diff
@@ -0,0 +1,159 @@
...
+			"rate": tax.get("rate") * 100.00,
+			"included_in_print_rate": set_included_in_print_rate(order)
+		})
+
+	taxes = update_taxes_with_shipping_rule(taxes, shipping_lines)
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py]: Should be `update_taxes_with_shipping_lines` as you are not using the Shipping Rule feature of ERPNext

File: erpnext_shopify/sync_orders.py [hunk 26]
```diff
@@ -0,0 +1,159 @@
...
+			"naming_series": shopify_settings.sales_order_series or "SO-Shopify-",
+			"shopify_id": order.get("id"),
+			"customer": frappe.db.get_value("Customer", {"shopify_id": order.get("customer").get("id")}, "name"),
+			"delivery_date": nowdate(),
+			"selling_price_list": shopify_settings.price_list,
+			"ignore_pricing_rule": 1,
+			"apply_discount_on": "Net Total",
+			"discount_amount": get_discounted_amount(order),
+			"items": get_item_line(order.get("line_items"), shopify_settings),
+			"taxes": get_tax_line(order, order.get("shipping_lines"), shopify_settings)
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py]: You can get `shipping_lines` from order itself in the get_tax_line function

File: erpnext_shopify/sync_orders.py [hunk 27]
```diff
@@ -0,0 +1,159 @@
...
+			"account_head": get_tax_account_head(shipping_charge),
+			"description": shipping_charge["title"],
+			"tax_amount": shipping_charge["price"]
+		})
+
+	return taxes
+
+def get_tax_account_head(tax):
+	tax_account =  frappe.db.get_value("Shopify Tax Account", \
+		{"parent": "Shopify Settings", "shopify_tax": tax.get("title")}, "tax_account")
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_orders.py:169]: Is it possible to search using some kind of tax id instead of tax title, as title could be changed?

Also do we sync taxes?

File: erpnext_shopify/sync_products.py [hunk 28]
```diff
@@ -0,0 +1,482 @@
...
+from frappe.utils import cstr, flt, nowdate, cint, get_files_path
+from .utils import disable_shopify_sync_for_item, disable_shopify_sync_on_exception
+from .shopify_requests import (get_request, post_request, get_shopify_items, put_request, 
+	get_shopify_item_image)
+
+shopify_variants_attr_list = ["option1", "option2", "option3"]
+
+def sync_products(price_list, warehouse):
+	sync_shopify_items(warehouse)
+	sync_erp_items(price_list, warehouse)
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_products.py]: sync_erpnext_items

File: erpnext_shopify/sync_products.py [hunk 29]
```diff
@@ -0,0 +1,482 @@
...
+
+def sync_shopify_items(warehouse):
+	for item in get_shopify_items():
+		make_item(warehouse, item)
+
+def make_item(warehouse, item):
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_products.py]: `shopify_item`

File: erpnext_shopify/sync_products.py [hunk 30]
```diff
@@ -0,0 +1,482 @@
...
+						"attribute_value": attr_value, 
+						"abbr": get_attribute_abbr(attr_value)
+					} 
+					for attr_value in attr.get("values")
+				]
+			}).insert()
+			attribute.append({"attribute": attr.get("name")})
+
+		else:
+			"check for attribute values"
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_products.py]: comments should start with `#`

File: erpnext_shopify/sync_products.py [hunk 31]
```diff
@@ -0,0 +1,482 @@
...
+	else:
+		item_details = frappe.db.get_value("Item", {"shopify_variant_id": item.get("id")},
+			["name", "stock_uom", "item_name"], as_dict=1)
+		if item_details:
+			name = item_details.name
+
+	return name, item_details
+
+def update_item(item_details, item_dict):
+	update_item = frappe.get_doc("Item", item_details.name)
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_products.py]: Avoid variable name same as the function name!

File: erpnext_shopify/sync_products.py [hunk 32]
```diff
@@ -0,0 +1,482 @@
...
+	# check if the item really exists on shopify
+	if item.get("shopify_id"):
+		try:
+			get_request("/admin/products/{}.json".format(item.get("shopify_id")))
+		except requests.exceptions.HTTPError, e:
+			if e.args[0] and e.args[0].startswith("404"):
+				disable_shopify_sync_for_item(erp_item)
+				return
+			else:
+				disable_shopify_sync_for_item(erp_item)
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_products.py]: why disable sync for any other HTTP error?

File: erpnext_shopify/sync_products.py [hunk 33]
```diff
@@ -0,0 +1,482 @@
...
+		if image_info["image"]:
+			try:
+				if not exist_item_image(item.shopify_id, image_info):
+					try:
+						post_request("/admin/products/{0}/images.json".format(item.shopify_id), image_info)
+					except requests.exceptions.HTTPError, e:
+						if e.args[0] and e.args[0].startswith("422"):
+							disable_shopify_sync_for_item(item)
+						else:
+							disable_shopify_sync_for_item(erp_item)
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_products.py]: why disable for any other HTTP error?

File: erpnext_shopify/sync_products.py [hunk 34]
```diff
@@ -0,0 +1,482 @@
...
+							raise
+						
+			except ShopifyError:
+				raise ShopifyError
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_products.py]: This doesn't make any sense. Catch and raise the same error?

File: erpnext_shopify/sync_products.py [hunk 35]
```diff
@@ -0,0 +1,482 @@
...
+				
+def exist_item_image(shopify_id, image_info):
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_products.py]: This function is poorly named and does not clearly state what it does. Also you should add comments inside the function to explain what these if conditions mean.

File: erpnext_shopify/sync_products.py [hunk 36]
```diff
@@ -0,0 +1,482 @@
...
+			attr_dict[attr.attribute].append(attr.attribute_value)
+
+			if attr.idx <= 3:
+				variant_list[i]["option"+cstr(attr.idx)] = attr.attribute_value
+
+		variant_item_code.append(item_variant.item_code)
+
+	for i, attr in enumerate(attr_dict):
+		options.append({
+            "name": attr,
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_products.py]: indentation issue

File: erpnext_shopify/sync_products.py [hunk 37]
```diff
@@ -0,0 +1,482 @@
...
+def update_item_stock_qty():
+	shopify_settings = frappe.get_doc("Shopify Settings", "Shopify Settings")
+	for item in frappe.get_all("Item", fields=['name', "item_code"], filters={"sync_with_shopify": 1, 
+		"disabled": ("!=", 1)}):
+		update_item_stock(item.item_code, shopify_settings)
+
+def update_item_stock(item_code, shopify_settings, bin=None):
+	item = frappe.get_doc("Item", item_code)
+	if item.sync_qty_with_shopify:
+		if not bin:
```
Comment (anandpdoshi) [inline erpnext_shopify/sync_products.py:428]: use `stock.utils.get_bin`

File: erpnext_shopify/webhooks.py [hunk 38]
```diff
@@ -0,0 +1,83 @@
+import frappe
+from frappe import _
+from functools import wraps
+import hashlib, base64, hmac, json
+from frappe.exceptions import AuthenticationError, ValidationError
+
```
Comment (anandpdoshi) [inline erpnext_shopify/webhooks.py:6]: I suppose this isn't used?

File: erpnext_shopify/sync_orders.py (hunk 27 above)
Comment (saurabh6790) [inline erpnext_shopify/sync_orders.py:169]: No we do not sync taxes. 

tax lines do not have Id 

```
"tax_lines": [
                {
                    "title": "VAT",
                    "price": "4375.00",
                    "rate": 0.125
                }
            ]
```
